tests/compilers/stack_limiter.rs

```diff
@@ -0,0 +1,238 @@
+"#;
+
+#[test]
+fn stack_limit_ok() {
+    let wat = OK_WAT;
+    let store = get_store();
+    let module = Module::new(&store, &wat).unwrap();
+    let instance = Instance::new_with_config(
+        &module,
+        unsafe { InstanceConfig::default().with_stack_limit(1000) },
+        &imports! {},
+    );
+    assert!(instance.is_ok());
+    let instance = instance.unwrap();
+    let main_func = instance
+        .exports
+        .get_function("main")
+        .expect("expected function main");
+    let e = main_func.call(&[]);
+    assert!(e.is_ok());
+}
+
+#[test]
+fn stack_limit_huge_limit() {
+    let wat = OK_WAT;
+    let store = get_store();
+    let module = Module::new(&store, &wat).unwrap();
+    let instance = Instance::new_with_config(
+        &module,
+        unsafe { InstanceConfig::default().with_stack_limit(0x7FFF_FFFF) },
+        &imports! {},
+    );
+    assert!(instance.is_ok());
+    let instance = instance.unwrap();
+    let main_func = instance
+        .exports
+        .get_function("main")
+        .expect("expected function main");
+    main_func.call(&[]).unwrap();
+}
+
+#[test]
+fn stack_limit_no_args() {
+    let wat = r#"
+        (func $foo
+            call $foo
+        )
+        (func (export "main")
+            call $foo
+        )
+    "#;
+
+    let store = get_store();
+    let module = Module::new(&store, &wat).unwrap();
+    let instance = Instance::new_with_config(
+        &module,
+        unsafe { InstanceConfig::default().with_stack_limit(1000) },
+        &imports! {},
+    );
+    assert!(instance.is_ok());
+    let instance = instance.unwrap();
+    let main_func = instance
+        .exports
+        .get_function("main")
+        .expect("expected function main");
+    match main_func.call(&[]) {
+        Err(err) => {
+            let trap = err.to_trap().unwrap();
+            assert_eq!(trap, TrapCode::StackOverflow);
+        }
+        _ => assert!(false),
+    }
+}
+
+#[test]
+fn deep_but_sane() {
+    let wat = r#"
+        (func $foo (param $p0 i32) (result i32)
+            local.get $p0
+            i32.const 1
+            i32.sub
+            local.set $p0
+            block $B0
+                local.get $p0
+                i32.const 0
+                i32.le_s
+                br_if $B0
+                local.get $p0
+                call $foo
+                drop
+            end
+            local.get $p0
+        )
+        (func (export "main")
+            i32.const 1000
+            call $foo
+            drop
+        )
+    "#;
+
+    let store = get_store();
+    let module = Module::new(&store, &wat).unwrap();
+    let instance = Instance::new_with_config(&module, InstanceConfig::default(), &imports! {});
+    assert!(instance.is_ok());
+    let instance = instance.unwrap();
+    let main_func = instance
+        .exports
+        .get_function("main")
+        .expect("expected function main");
+
+    let e = main_func.call(&[]);
+    assert!(e.is_ok());
```

Comment:
Let's do it in a `for _ in 0..10` loop to check the `.do_call` machinery (that is, that we do reset the counter on each entry).

Reply:
Good idea, fixed.